### PR TITLE
Autogenerate pipeline from json config file

### PR DESCRIPTION
**Pull request by catalindumitru**
Instead of having a static pipeline.yml, autogenerate it from a JSON
configuration file. The json approach is taken because we also want to
be able to run tests locally, and thus the json can be parsed once for
generating the pipeline, and once for generating a local test run.

### CODE REVIEW

File: .buildkite/test_description.json
Comment (lauralt) [inline .buildkite/test_description.json:53]: Is this still needed?

File: .buildkite/autogenerate_pipeline.py
Comment (lauralt) [inline .buildkite/autogenerate_pipeline.py:214]: Can we leave a comment explaining what __dict__ does?

File: .buildkite/test_description.json
Comment (catalindumitru) [reply .buildkite/test_description.json]: Removed the platform key for the tests that don't need it.

File: .buildkite/autogenerate_pipeline.py
Comment (catalindumitru) [reply .buildkite/autogenerate_pipeline.py]: Done. Also replaced `self.__dict__` with `vars(self)`. This seems to be the more "Pythonic" way because it avoids calling magic methods such as `__dict__`.

File: .buildkite/autogenerate_pipeline.py
Comment (andreeaflorescu) [inline .buildkite/autogenerate_pipeline.py:46]: We should add a comment about this const to say that it represents the rust-vmm-container used for running the tests.

Same for the other const, we should say that it represents the Buildkite docker plugin version.

File: .buildkite/autogenerate_pipeline.py
Comment (andreeaflorescu) [inline .buildkite/autogenerate_pipeline.py:68]: If I remember correctly, the order in which we define these matters. We should add a comment in the code here to make sure that people don't change the order by mistake.

File: .buildkite/autogenerate_pipeline.py
Comment (andreeaflorescu) [inline .buildkite/autogenerate_pipeline.py:85]: Can you add a comment here why you need to do this? (it's a problem for the way we are setting the tag on the host, it should've been aarch64; nevertheless we should explain it)

File: .buildkite/autogenerate_pipeline.py
Comment (andreeaflorescu) [inline .buildkite/autogenerate_pipeline.py:137]: Can you add a comment about the logic here? It only becomes clear what you want to do when reading the build function. We should say that when no platform is specified, we don;t add the platform to the dictionary so that the test can run in any env.

**Review by andreeaflorescu — COMMENTED** [review]
LGTM high level. We should add some information in the readme about how this works now, and why we ended up implementing it this way.

File: .buildkite/autogenerate_pipeline.py
Comment (catalindumitru) [reply .buildkite/autogenerate_pipeline.py:48]: Done.

File: .buildkite/autogenerate_pipeline.py
Comment (catalindumitru) [reply .buildkite/autogenerate_pipeline.py]: Done.

File: .buildkite/autogenerate_pipeline.py
Comment (catalindumitru) [reply .buildkite/autogenerate_pipeline.py:92]: Done.

File: .buildkite/autogenerate_pipeline.py
Comment (catalindumitru) [reply .buildkite/autogenerate_pipeline.py:148]: Done.

**Review by andreeaflorescu — DISMISSED** [review]

File: README.md
Comment (andreeaflorescu) [inline README.md:84]: nit: should this be on the following line?

File: README.md
Comment (andreeaflorescu) [inline README.md:63]: Let's also write about the defaults and say that for most use cases overriding these is not necessary.

We might also want to give a few examples when we do want to override them (platform needs a custom device that is not available on the existing test instances, need for a specialized hypervisor - for example, microsoft hyper-v).

File: README.md
Comment (andreeaflorescu) [inline README.md:52]: Can you add a link to the process from the community repo? Here is the link: https://github.com/rust-vmm/community/blob/master/docs/maintainers/setup_new_repo.md#set-up-ci

File: README.md
Comment (catalindumitru) [reply README.md]: Yes, changed it.

File: README.md
Comment (catalindumitru) [reply README.md:64]: Done.

File: README.md
Comment (catalindumitru) [reply README.md:52]: Added the link.

**Review by andreeaflorescu — DISMISSED** [review]

File: README.md
Comment (lauralt) [inline README.md:60]: I would update at lest the example from #Buildkite Pipeline section with how it will look after running autogenerate_pipeline.py? (fewer newlines etc.).

**Review by lauralt — COMMENTED** [review]
Looks good! I'll approve once we manage to test this end to end.

File: README.md
Comment (catalindumitru) [reply README.md:60]: Updated the other examples.

**Review by lauralt — APPROVED** [review]

**Review by andreeaflorescu — APPROVED** [review]

**Pull request merged**
